pallets/liquidity-pools/src/message.rs

```diff
@@ -66,6 +72,121 @@ impl TryInto<Domain> for SerializableDomain {
 	}
 }
 
+/// A message type that can not be a batch.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct NoBatchMessage(Message);
+
+impl TryFrom<Message> for NoBatchMessage {
+	type Error = DispatchError;
+
+	fn try_from(message: Message) -> Result<Self, DispatchError> {
+		match message {
+			Message::Batch { .. } => Err(DispatchError::Other("A submessage can not be a batch")),
+			_ => Ok(Self(message)),
+		}
+	}
+}
+
+impl MaxEncodedLen for NoBatchMessage {
+	fn max_encoded_len() -> usize {
+		// This message use a non batch message version to obtain the encoded
+		// len to avoid an infite recursion: message -> batch -> message -> batch ...
+		Message::<()>::max_encoded_len()
+	}
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo, MaxEncodedLen, Default)]
+pub struct BatchMessages(BoundedVec<NoBatchMessage, ConstU32<MAX_BATCH_MESSAGES>>);
+
+impl Serialize for BatchMessages {
+	fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+		// Serializing as a tuple to avoid the prefix length used for dynamic lists
+		let mut tuple = serializer.serialize_tuple(self.0.len())?;
+		for msg in self.0.iter() {
+			let encoded = gmpf::to_vec(&msg.0).map_err(|e| S::Error::custom(e.to_string()))?;
+
+			// Serializing as bytes automatically encodes the prefix size
+			tuple.serialize_element(&encoded)?;
+		}
+		tuple.end()
+	}
+}
+
+impl<'de> Deserialize<'de> for BatchMessages {
+	fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+		// We need a custom visitor because we do not know the length upfront
+		struct MsgVisitor;
+
+		impl<'de> Visitor<'de> for MsgVisitor {
+			type Value = BatchMessages;
+
+			fn expecting(&self, formatter: &mut sp_std::fmt::Formatter) -> sp_std::fmt::Result {
+				formatter.write_str("A sequence of pairs size-submessage")
+			}
+
+			fn visit_seq<A: SeqAccess<'de>>(self, mut seq: A) -> Result<Self::Value, A::Error> {
+				let mut batch = BatchMessages::default();
+
+				// We only stop on error trying to deserialize the length of the submessage.
+				// The error will happen when we reach EOF
+				while seq.next_element::<u16>().unwrap_or(None).is_some() {
+					let msg = seq
+						.next_element()?
+						.ok_or(A::Error::custom("expected submessage"))?;
+
+					batch
+						.try_add(msg)
+						.map_err(|e| A::Error::custom::<&'static str>(e.into()))?;
+				}
+
+				Ok(batch)
+			}
+		}
+
+		let limit = MAX_BATCH_MESSAGES as usize * 2; // Lengths and messages
+		deserializer.deserialize_tuple(limit, MsgVisitor)
```

Comment:
Shouldn't we for deserialze also take care of handling the `BoundedVec` or will the internal scale-codec handle that?

Reply:
The `BatchMessages` knows how to create itself by increasing the internal `BoundedVec` but we can not use the substrate `BoundedVec` deserialization because any variable list of elements requires known the size upfront but we can not.

Basically, we need to redo this process due to the solidity requirement of not sending the batch size upfront.